lib/CFRPageActions.jsm

```diff
@@ -275,6 +276,52 @@ class PageAction {
     }
   }
 
+  async _renderPinTabAnimation() {
+    const footer = this.window.document.getElementById("cfr-notification-footer");
+    const ANIMATION_CONTAINER_ID = "cfr-notification-footer-pintab-animation-container";
+    let animationsEnabled = Services.prefs.getBoolPref(ANIMATIONS_ENABLED_PREF, true);
+    let animationContainer = this.window.document.getElementById(ANIMATION_CONTAINER_ID);
+    if (animationContainer) {
+      animationContainer.toggleAttribute("animate", animationsEnabled);
+      animationContainer.removeAttribute("paused");
+    } else {
+      animationContainer = this.window.document.createElement("vbox");
+      animationContainer.setAttribute("id", ANIMATION_CONTAINER_ID);
```

Comment:
I see a pattern of `.createElement(); .setAttribute("id",....)` that I guess could be factored out?

Reply:
along with the `.appendChild()` perhaps. 🙄

Reply:
something like this ?

```javascript
  _createElement(type, parent, id) {
    let element = this.window.document.createElement(type);
    if (id) {
      element.setAttribute("id", id);
    }
    parent.appendChild(element);
    return element;
  }

  async _renderPinTabAnimation() {
    const footer = this.window.document.getElementById("cfr-notification-footer");
    const ANIMATION_CONTAINER_ID = "cfr-notification-footer-pintab-animation-container";
    let animationsEnabled = Services.prefs.getBoolPref(ANIMATIONS_ENABLED_PREF, true);
    let animationContainer = this.window.document.getElementById(ANIMATION_CONTAINER_ID);
    if (animationContainer) {
      animationContainer.toggleAttribute("animate", animationsEnabled);
      animationContainer.removeAttribute("paused");
    } else {
      animationContainer = this._createElement("vbox", footer, ANIMATION_CONTAINER_ID);
      animationContainer.toggleAttribute("animate", animationsEnabled);

      // spacer
      this._createElement("vbox", animationContainer).setAttribute("flex", 1);

      let controlsContainer = this._createElement("hbox", animationContainer, "cfr-notification-footer-animation-controls");

      // spacer
      this._createElement("vbox", controlsContainer).setAttribute("flex", 1);

      let pauseButton = this._createElement("hbox", controlsContainer, "cfr-notification-footer-pause-button");
      pauseButton.addEventListener("click", () => {
        animationContainer.setAttribute("paused", true);
      });

      let pauseLabel = this._createElement("label", pauseButton, "cfr-notification-footer-pause-label");
      pauseLabel.textContent = await this.getStrings({"string_id": "cfr-doorhanger-pintab-animation-pause"});

      // pause icon
      this._createElement("image", pauseButton, "cfr-notification-footer-pause-icon");
    }
  }
```

Reply:
I like it. Writing unit tests that mess with the DOM is weird, this should make it nicer.

Reply:
I would group `type` and `id` into an options object and have something like
```
_createElementAndAppend({id, type}, parent)
```